Rails importmap lesson: Update to add more beginner friendly terminology (#23678)
TheOdinProject/curriculum: rails_programming/rails_basics/importmap.md

### Introduction

> Prior to Rails 7, there was a long running issue with how Javascript third-party packages should be managed in a Rails application. The original approach was to wrap the Javascript in a Ruby Gem and release it. While this did bring about versioning and stability it also made it slow to update to the latest releases of libraries as you needed to wait for the maintainers of the Ruby Gem to update it, test it, and release a new version of the Gem. Rails 6 tried to fix this solution by creating a wrapper Gem around [webpack](https://webpack.js.org/). The gem, called Webpacker, used Rails' famous convention over configuration to give the ability to use the latest JS libraries but with some sensible defaults. This wasn't without issue though, the main one being that you needed to have a deeper understanding of Webpacker if you wanted to move away from those conventions. Unfortunately it didn't solve the issues it was meant to and a new approach was needed.

Reviewer: ```suggestion
Prior to Rails 7, there was a long running issue with how Javascript third-party packages should be managed in a Rails application. The original approach was to wrap the Javascript in a Ruby Gem and release it. While this did bring about versioning and stability, it also made it slow to update to the latest releases of libraries as you needed to wait for the maintainers of the Ruby Gem to update it, test it, and release a new version of the gem.

Rails 6 tried to fix this solution by creating a wrapper gem around [webpack](https://webpack.js.org/). The gem, called Webpacker, used Rails' famous convention over configuration to give the ability to use the latest JavaScript libraries but with some sensible defaults. This wasn't without issue though, the main one being that you needed to have a deeper understanding of Webpacker if you wanted to move away from those conventions. Unfortunately it didn't solve the issues it was meant to and a new approach was needed.
```

With Rails 7 we have [importmap](https://github.com/rails/importmap-rails). This approach to managing JavaScript in your applications was chosen for a couple of reasons. First ES6 is now supported by all browsers that matter. Chrome, Edge, Safari, and Firefox. Secondly, HTTP2 is now the norm meaning you can have a single connection that responds to the client with multiple small files. Now you can get rid of your bundler for JavaScript.

### Learning Outcomes

*Look through these now and then use them to test yourself after doing the assignment*

> *What are Import maps?

Reviewer: ```suggestion
* What are Import maps?
```

> *How do Import maps work?

Reviewer: ```suggestion
* How do Import maps work?
```

> *How do you use Import maps?

Reviewer: ```suggestion
* How do you use Import maps?
```

*What are `pin`s?

### Import Maps

Import maps let you import JavaScript modules using logical names that map to versioned/digested files – directly from the browser. So you can build modern JavaScript applications using JavaScript libraries made for ES modules (ESM) without the need for [transpiling](https://en.wikipedia.org/wiki/Source-to-source_compiler) or [bundling](https://www.dottedsquirrel.com/bundling-javascript/). This frees you from needing Webpack, Yarn, npm, or any other part of the JavaScript toolchain. All you need is the asset pipeline that's already included in Rails.

With this approach, you'll ship many small JavaScript files instead of one big JavaScript file. Thanks to HTTP/2 that no longer carries a material performance penalty during the initial transport, and in fact offers substantial benefits over the long run due to better caching dynamics. Whereas before any change to any JavaScript file included in your big bundle would invalidate the cache for the whole bundle, now only the cache for that single file is invalidated.

#### How do import maps work?

> The first thing to know is that import maps are not a rails specific solution. They are an existing solution that the Rails team thought would be the best way to bring Javascript to Rails applications as a sane default. The basic premise is that you can import Javascript libraries into your application just referencing them by name rather than having to reference the exact location of where the library can be found. Import maps are essentially string substitution for "bare module specifiers". If you're not familiar with "bare module specifiers" they look like this: `import React from "react"`. However just writing imports this way isn't something currently valid as a way to to load in Javascript code to the current file. The valid ways to do this currently are defined under something called the ESM loader spec which if you really want to you can read more about [here](https://nodejs.org/api/esm.html)

Reviewer: ```suggestion
The first thing to know is that import maps are not a Rails specific solution. They are an existing solution that the Rails team thought would be the best way to bring Javascript to Rails applications as a sane default. The basic premise is that you can import Javascript libraries into your application just referencing them by name rather than having to reference the exact location of where the library can be found. Import maps are essentially string substitution for "bare module specifiers". If you're not familiar with "bare module specifiers" they look like this: `import React from "react"`. However just writing imports this way isn't something currently valid as a way to to load in Javascript code to the current file. The valid ways to do this currently are defined under something called the ESM loader spec which if you really want to you can read more about from the [Node.js docs](https://nodejs.org/api/esm.html).
```

* HTTP path:
    `import React from "https://ga.jspm.io/npm:react@17.0.1/index.js"`

The `importmap-rails` gem provides an API for mapping the "bare module specifiers" to 1 of the 3 ways of loading ESM JavaScript packages.

Example:

~~~Ruby
# config/importmap.rb
pin "react", to: "https://ga.jspm.io/npm:react@17.0.2/index.js"
~~~

> This means "every time you see `import React from "react"` in the code it will actually be like writing `import React from "https://ga.jspm.io/npm:react@17.0.2/index.js"`"

Reviewer: ```suggestion
This means that every time you see `import React from "react"` in the code it will actually be like writing `import React from "https://ga.jspm.io/npm:react@17.0.2/index.js"`.
```

#### Using Import Maps

> When creating a new Rails 7+ application `importmap-rails` will be automatically added to your Gemfile. `importmap-rails` is doing a few things for you. First the `<%= javascript_importmap_tags %>` line is added to `views/layouts/application.html.erb` which will setup the JSON configuration inside of a `<script type="importmap">` tag. Then the [es-module-shim](https://github.com/guybedford/es-module-shims) is loaded. Finally, the application entry point is imported via `<script type="module">import "application"</script>`. Inside your `app/javascript/application.js` you set up the application by importing any modules that have been defined in the import map. It is good to use logical names that match the package names used by NPM, that way if you want to start transpiling or bundling your code you won't have to change any module imports.

Reviewer: ```suggestion
When creating a new Rails 7+ application `importmap-rails` will be automatically added to your Gemfile. `importmap-rails` is doing a few things for you. First the `<%= javascript_importmap_tags %>` line is added to `views/layouts/application.html.erb` which will set up the JSON configuration inside of a `<script type="importmap">` tag. Then the [es-module-shim](https://github.com/guybedford/es-module-shims) is loaded. Finally, the application entry point is imported via `<script type="module">import "application"</script>`. Inside your `app/javascript/application.js` you set up the application by importing any modules that have been defined in the import map. It is good to use logical names that match the package names used by NPM, that way if you want to start transpiling or bundling your code you won't have to change any module imports.
```

You start the setup via the configuration file located `config/importmap.rb`. This file is automatically reloaded in development when anything changes. That being said, if you remove pins and need them gone from the rendered importmap or list of preloads you MUST restart the server.

### Import Maps with NPM packages

`importmap-rails` is designed to be used with JavaScript CDNs for NPM package dependencies.

You can use the `./bin/importmap` command that's added as part of the install to `pin`, `unpin`, or `update` npm packages in your import map. This command resolves the package dependencies and adds the pins to your `config/importmap.rb`.

~~~JS
./bin/importmap pin react react-dom

> Pinning "react" to https://ga.jspm.io/npm:react@17.0.2/index.js

Reviewer 2: I know it's not part of this PR, but I think using a package the student can follow and code along with would be better for the code examples. Perhaps something like lodash? they can pin it and use one the lodash string functions to console.log something. I'm a little anxious of these examples going over students heads if they can't see it working for themselves.

But that could be a follow up if you think it's worth doing? it doesn't have to stop this from going out.

Author: @KevinMulhern That's a good point mate. Usually I wouldn't mind reworking it but I think this could be a wider thing we look into as part of the rails_path ownership where we ensure all examples are runnable in some way for those that want to try.

    "scheduler": "https://ga.jspm.io/npm:scheduler@0.20.2/index.js"
  }
}
~~~

The two dependencies `react` and `react-dom` resolve to four total dependencies. Now you can import them in your Javascript code.

~~~JS
import React from "react"
import ReactDOM from "react-dom"
~~~

Using the `./bin/importmap` is just a convenience wrapper for resolving logical package names to CDN URLs. You can just look up the CDN URLs yourself and pin those in `config/importmap.rb`:

~~~Ruby
pin "react", to: "https://cdn.skypack.dev/react"
~~~

### Downloading vendor files

> If you don't want to use a CDN in production, you can download vendored files from the CDN when you're setting up your pins. These packages are downloaded to your `vendor/javascript:

Reviewer: ```suggestion
If you don't want to use a CDN in production, you can download vendored files from the CDN when you're setting up your pins. These packages are downloaded to your `vendor/javascript`:
```
